content/docs/guidelines/best-practices/godot-gdscript/_index.md

```diff
@@ -0,0 +1,369 @@
++++
+author = "nathan"
+date = "2019-07-16T08:34:54+09:00"
+description = "This guide covers some best practices to write solid GDScript code, to keep a sane code-base when developing projects of growing sizes."
+title = "Best Practices: Godot GDScript"
+menuTitle = "Godot GDScript"
+weight = 2
+aliases = ["/open-source/guidelines/godot-gdscript/"]
+
+[[resources]]
+  name = "banner"
+  src = "banner.jpg"
+
++++
+
+This guide covers some best practices to write solid GDScript code. It's the style we use to keep our code clean and maintainable, working on growing Godot games. It's meant as a complementary resource to the [oficial GDScript guidelines](http://docs.godotengine.org/en/3.2/getting_started/scripting/gdscript/gdscript_styleguide.html).
+
+The ideas below draw inspiration from good practices from different paradigms and languages. Especially from the work of the Python community, some functional programming principles, and the official GDScript documentation:
+
+1. [GDScript Style Guide](//docs.godotengine.org/en/latest/getting_started/scripting/gdscript/gdscript_styleguide.html)
+1. [Static typing in GDScript](//docs.godotengine.org/en/latest/getting_started/scripting/gdscript/static_typing.html)
+1. [Docs writing guidelines](//docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html)
+1. [Boundaries - A talk by Gary Bernhardt from SCNA 2012](//www.destroyallsoftware.com/talks/boundaries) & [Functional Core, Imperative Shell](//www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)
+1. [The Clean Architecture in Python](//www.youtube.com/watch?v=DJtef410XaM)
+1. [Onion Architecture Without the Tears - Brendan Richards](//www.youtube.com/watch?v=R2pW09tMCnE&t=1095s)
+1. [Domain Driven Design Through Onion Architecture](//www.youtube.com/watch?v=pL9XeNjy_z4)
+
+Godot is mostly object-oriented and offers its own tools to make objects communicate together, like signals, and the node tree. As a result, it's not always easy to apply principles and techniques from other programming languages to it.
+
+This guideline is created with a few goals in mind:
+
+- Avoiding coupling, having system depend on and break one another
+- Managing boundaries, the places where different game systems interact with one another
+- Keeping the code readable working as a team, as we developers spend more time reading than writing code
+
+## Code Writing Style
+
+{{< youtube FB2sqJgfqXI >}}
+
+We use the same code style so that you can find the information you need fast.
+
+### In Short
+
+We follow the same [code order](http://docs.godotengine.org/en/3.2/getting_started/scripting/gdscript/gdscript_styleguide.html#code-order) from the official GDScript style guide. Here is a complete example that follows the guidelines below:
+
+{{< highlight gdscript >}}
+class_name StateMachine
+extends Node
+# Hierarchical State machine for the player.
```

Comment:
The docstrings should now be above class_name, to support the class reference system and Godot's language server docs feature